[rom_ext] Create bazel helpers for releasing imm_section (#26735)
lowRISC/opentitan: sw/device/silicon_creator/rom_ext/imm_section/README.md

# RomExt Immutable Section Release Guide

## Bump the release version

> Edit the `IMM_SECTION_VERSION_{MAJOR,MINOR}"` line in the bazel file.

Reviewer: After editing, a commit and PR should be made so that one builds from a clean tree.

Building / stamping from a dirty tree should only be done for testing purposes.

`sw/device/silicon_creator/rom_ext/imm_section/defs.bzl`

## Build a release bundle

```shell
$ bazel build --stamp //sw/device/silicon_creator/rom_ext/imm_section:release_bundle
```

Building with --stamp flag will stamp the built immutable section, including

> the commit hash and the workspace status.

Reviewer: add a note here saying that we only want to release an immutable section built from a clean workspace?

The command will create a release tarball in
`bazel-bin/sw/device/silicon_creator/rom_ext/imm_section/release_bundle.tar`.

> The built tarball contains the:

Reviewer: ```suggestion
The built tarball contains:
```

> *   A generated BUILD file that can be referenced by the ROM\_EXT target

Reviewer: ```suggestion
*   a generated BUILD file that can be referenced by the ROM\_EXT target.
```

> *   Compiled imm\_section binary for each variant

Reviewer: ```suggestion
*   compiled imm\_section binaries for each variant, and
```

> *   Elf / map files for troubleshooting

Reviewer: ```suggestion
*   elf / map files corresponding to each binary for debugging purposes.
```

## Submit a prebuilt release

Extracting the tarball to your prebuilt folder will create the prebuilt targets
that can be linked to ROM\_EXT. The prebuilt folder could be either:

> *   `sw/device/silicon_creator/rom\_ext/imm_section/prebuilts/`
> *   Folder in the SKU extension repo.

Reviewer: ```suggestion
*   this (upstream) repo under: `sw/device/silicon_creator/rom\_ext/imm_section/prebuilts/`, or
*   to a (downstream) integrators repo.
```

$ ./bazelisk.sh query //sw/device/silicon_creator/rom_ext/imm_section/prebuilts/2025-03-28-v0.1-d686b36b/...
//sw/device/silicon_creator/rom_ext/imm_section/prebuilts/2025-03-28-v0.1-d686b36b:dice_cwt
//sw/device/silicon_creator/rom_ext/imm_section/prebuilts/2025-03-28-v0.1-d686b36b:dice_cwt_object
//sw/device/silicon_creator/rom_ext/imm_section/prebuilts/2025-03-28-v0.1-d686b36b:dice_cwt_otp
//sw/device/silicon_creator/rom_ext/imm_section/prebuilts/2025-03-28-v0.1-d686b36b:dice_x509
//sw/device/silicon_creator/rom_ext/imm_section/prebuilts/2025-03-28-v0.1-d686b36b:dice_x509_object
//sw/device/silicon_creator/rom_ext/imm_section/prebuilts/2025-03-28-v0.1-d686b36b:dice_x509_otp
```

## Create ROM\_EXT with a prebuilt release

The immutable section is linked to ROM\_EXT by deps. To build a ROM\_EXT with
prebuilt release, please update the deps to the corresponding variants listed
above. For example,

```diff
  deps = [
-     "@//sw/device/silicon_creator/rom_ext/imm_section:main_section_dice_cwt_slot_virtual",
+     "@//sw/device/silicon_creator/rom_ext/imm_section/prebuilts/2025-03-28-v0.1-d686b36b:dice_cwt",
  ]
```

> ## Enforce immutability hash check with a prebuilt release

Reviewer: ```suggestion
## Enforce immutability hash check in OTP
```

# RomExt Immutable Section Release Guide

## Bump the release version

Edit the `IMM_SECTION_VERSION = "0.1"` line in the bazel file.
`sw/device/silicon_creator/rom_ext/imm_section/defs.bzl`

## Build a release bundle

```shell
$ bazel build //sw/device/silicon_creator/rom_ext/imm_section:release_bundle
```

The command will create a release tarball in
`bazel-bin/sw/device/silicon_creator/rom_ext/imm_section/release_bundle.tar`.

The built tarball contains the:

*   A generated BUILD file that can be referenced by the ROM\_EXT target
*   Compiled imm\_section binaries

> *   OTP overlay to enable immutability hash enforcement

Reviewer: Lets remove this from the bundle. This should just contain a single binary per immutable_rom_ext variation, since the immutable section is the same across exec_envs (preferably the silicon_creator), and the map/elf files too if you want.

Author: Would you mind me including the auto-generated BUILD file?

Reviewer: Sure, the build file is convenient.

v0.1/release_bundle_dice_cwt_otp.with_immutable_re_params.json
v0.1/release_bundle_dice_x509_otp.with_immutable_re_params.json
v0.1/version.txt

$ ./bazelisk.sh query //sw/device/silicon_creator/rom_ext/imm_section/prebuilts/v0.1/...
//sw/device/silicon_creator/rom_ext/imm_section/prebuilts:nop_imm_section
//sw/device/silicon_creator/rom_ext/imm_section/prebuilts/v0.1:dice_cwt
//sw/device/silicon_creator/rom_ext/imm_section/prebuilts/v0.1:dice_cwt_object
//sw/device/silicon_creator/rom_ext/imm_section/prebuilts/v0.1:dice_cwt_otp
//sw/device/silicon_creator/rom_ext/imm_section/prebuilts/v0.1:dice_x509
//sw/device/silicon_creator/rom_ext/imm_section/prebuilts/v0.1:dice_x509_object
//sw/device/silicon_creator/rom_ext/imm_section/prebuilts/v0.1:dice_x509_otp
```

## Create ROM\_EXT with a prebuilt release

The immutable section is linked to ROM\_EXT by deps. To build a ROM\_EXT with
prebuilt release, please update the deps to the corresponding variants listed
above. For example,

```diff
  deps = [
-     "@//sw/device/silicon_creator/rom_ext/imm_section:main_section_dice_cwt_slot_virtual",
+     "@//sw/device/silicon_creator/rom_ext/imm_section/prebuilts/v0.1:dice_cwt",
  ]
```

> ## Enforce immutability hash check with a prebuilt release

Reviewer: we don't need to check in the OTP overlay, you can just have a pointer in this readme to an exiting example OTP overlay

Reviewer: We dont need any otp_json_immutable_rom_ext rule. See my example build downstream.

Reviewer: However, I think you should have a section in this readme saying that an integrator needs to flip some bits in their SKU's OTP image to enforce the hash of the immutable section.

Author: I have dropped OTP related commits, and added an overlay example in README.
